// node-pg-migrate/dist/db.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pg = require('pg');

var _pg2 = _interopRequireDefault(_pg);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// or native libpq bindings
// import pg from 'pg/native';

exports.default = function (connection_string) {
  var client = new _pg2.default.Client(connection_string);
  var client_active = false;
  var beforeCloseListeners = [];

  var createConnection = function createConnection() {
    return new Promise(function (resolve, reject) {
      return client_active ? resolve() : client.connect(function (err) {
        if (err) {
          console.error('could not connect to postgres', err);
          reject(err);
        } else {
          client_active = true;
          resolve();
        }
      });
    });
  };

  var query = function query(string) {
    return createConnection().then(function () {
      return new Promise(function (resolve, reject) {
        return client.query(string, function (err, result) {
          return err ? reject(err) : resolve(result);
        });
      });
    }).catch(function (err) {
      var message = err.message,
          position = err.position;

      if (message && position >= 1) {
        var endLineWrapIndexOf = string.indexOf('\n', position);
        var endLineWrapPos = endLineWrapIndexOf >= 0 ? endLineWrapIndexOf : string.length;
        var stringStart = string.substring(0, endLineWrapPos);
        var stringEnd = string.substr(endLineWrapPos);
        var startLineWrapPos = stringStart.lastIndexOf('\n') + 1;
        var padding = ' '.repeat(position - startLineWrapPos - 1);
        console.error(`Error executing:
${stringStart}
${padding}^^^^${stringEnd}

${message}
`);
      } else {
        console.error(`Error executing:
${string}
${err}
`);
      }
      throw err;
    });
  };

  var select = function select(string) {
    return query(string).then(function (result) {
      return result && result.rows;
    });
  };
  var column = function column(string, column_name) {
    return select(string).then(function (rows) {
      return rows.map(function (r) {
        return r[column_name];
      });
    });
  };

  return {
    query,
    select,
    column,

    addBeforeCloseListener: function addBeforeCloseListener(listener) {
      return beforeCloseListeners.push(listener);
    },

    close: function close() {
      return beforeCloseListeners.reduce(function (promise, listener) {
        return promise.then(listener).catch(function (err) {
          return console.err(err.stack ? err.stack : err);
        });
      }, Promise.resolve()).then(function () {
        client_active = false;
        if (client) {
          client.end();
        }
      });
    }
  };
}; /*
    This file just manages the database connection and provides a query method
    */
