// node-pg-migrate/dist/migration-builder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      The migration builder is used to actually create a migration from instructions
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      A new instance of MigrationBuilder is instantiated and passed to the up or down block
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      of each migration when it is being run.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      It makes the methods available via the pgm variable and stores up the sql commands.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      This is what makes it possible to do this without making everything async
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      and it makes inferrence of down migrations possible.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */

var _utils = require('./utils');

var _extensions = require('./operations/extensions');

var extensions = _interopRequireWildcard(_extensions);

var _indexes = require('./operations/indexes');

var indexes = _interopRequireWildcard(_indexes);

var _tables = require('./operations/tables');

var tables = _interopRequireWildcard(_tables);

var _other = require('./operations/other');

var other = _interopRequireWildcard(_other);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MigrationBuilder = function () {
  function MigrationBuilder() {
    var _this = this;

    _classCallCheck(this, MigrationBuilder);

    this._steps = [];
    this._REVERSE_MODE = false;
    // by default, all migrations are wrapped in a transaction
    this._use_transaction = true;

    // this function wraps each operation within a function that either
    // calls the operation or its reverse, and appends the result (array of sql statements)
    // to the  steps array
    var wrap = function wrap(operation) {
      return function () {
        if (_this._REVERSE_MODE && typeof operation.reverse !== 'function') {
          throw new Error('Impossible to automatically infer down migration');
        }
        _this._steps = _this._steps.concat(_this._REVERSE_MODE ? operation.reverse.apply(operation, arguments) : operation.apply(undefined, arguments));
      };
    };

    // defines the methods that are accessible via pgm in each migrations
    // there are some convenience aliases to make usage easier
    this.createExtension = wrap(extensions.create);
    this.dropExtension = wrap(extensions.drop);
    this.addExtension = this.createExtension;

    this.createTable = wrap(tables.create);
    this.dropTable = wrap(tables.drop);
    this.renameTable = wrap(tables.renameTable);

    this.addColumns = wrap(tables.addColumns);
    this.dropColumns = wrap(tables.dropColumns);
    this.renameColumn = wrap(tables.renameColumn);
    this.alterColumn = wrap(tables.alterColumn);
    this.addColumn = this.addColumns;
    this.dropColumn = this.dropColumns;

    this.addConstraint = wrap(tables.addConstraint);
    this.dropConstraint = wrap(tables.dropConstraint);
    this.createConstraint = this.addConstraint;

    this.createType = wrap(tables.createType);
    this.dropType = wrap(tables.dropType);
    this.alterType = wrap(tables.alterType);
    this.addType = this.createType;

    this.createIndex = wrap(indexes.create);
    this.dropIndex = wrap(indexes.drop);
    this.addIndex = this.createIndex;

    this.sql = wrap(other.sql);

    // Other utilities which may be useful
    // .func creates a string which will not be escaped
    // common uses are for PG functions, ex: { ... default: pgm.func('NOW()') }
    this.func = _utils.PgLiteral.create;
  }

  _createClass(MigrationBuilder, [{
    key: 'enableReverseMode',
    value: function enableReverseMode() {
      this._REVERSE_MODE = true;
      return this;
    }
  }, {
    key: 'noTransaction',
    value: function noTransaction() {
      this._use_transaction = false;
      return this;
    }
  }, {
    key: 'isUsingTransaction',
    value: function isUsingTransaction() {
      return this._use_transaction;
    }
  }, {
    key: 'getSql',
    value: function getSql() {
      return `${this.getSqlSteps().join('\n')}\n`;
    }
  }, {
    key: 'getSqlSteps',
    value: function getSqlSteps() {
      // in reverse mode, we flip the order of the statements
      return this._REVERSE_MODE ? this._steps.slice().reverse() : this._steps;
    }
  }]);

  return MigrationBuilder;
}();

exports.default = MigrationBuilder;
