// node-pg-migrate/dist/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// This is used to create unescaped strings
// exposed in the migrations via pgm.func
var PgLiteral = exports.PgLiteral = function () {
  _createClass(PgLiteral, null, [{
    key: 'create',
    value: function create(str) {
      return new PgLiteral(str);
    }
  }]);

  function PgLiteral(str) {
    _classCallCheck(this, PgLiteral);

    this._str = str;
  }

  _createClass(PgLiteral, [{
    key: 'toString',
    value: function toString() {
      return this._str;
    }
  }]);

  return PgLiteral;
}();

var schemalize = function schemalize(v) {
  if (typeof v === 'object') {
    var schema = v.schema;
    var name = v.name;
    return (schema ? `${schema}"."` : '') + name;
  }
  return v;
};

var t = exports.t = function t(s, d) {
  return Object.keys(d || {}).reduce(function (str, p) {
    return str.replace(new RegExp(`{${p}}`, 'g'), schemalize(d[p]));
  }, s);
};

var escapeValue = exports.escapeValue = function escapeValue(val) {
  if (val === null) {
    return 'NULL';
  }
  if (typeof val === 'boolean') {
    return val.toString();
  }
  if (typeof val === 'string') {
    return `'${escape(val)}'`;
  }
  if (typeof val === 'number') {
    return val;
  }
  if (Array.isArray(val)) {
    return `ARRAY[${val.map(escapeValue).join(',').replace(/ARRAY/g, '')}]`;
  }
  if (val instanceof PgLiteral) {
    return val.toString();
  }
  return '';
};

var template = exports.template = function template(strings) {
  for (var _len = arguments.length, keys = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    keys[_key - 1] = arguments[_key];
  }

  var result = [strings[0]];
  keys.forEach(function (key, i) {
    result.push(schemalize(key), strings[i + 1]);
  });
  return result.join('');
};

var getMigrationTableSchema = exports.getMigrationTableSchema = function getMigrationTableSchema(options) {
  return options.migrations_schema || options.schema || 'public';
};

var finallyPromise = exports.finallyPromise = function finallyPromise(func) {
  return [func, function (err) {
    var errHandler = function errHandler(innerErr) {
      console.err(innerErr.stack ? innerErr.stack : innerErr);
      throw err;
    };
    try {
      return Promise.resolve(func()).then(function () {
        throw err;
      }, errHandler);
    } catch (innerErr) {
      return errHandler(innerErr);
    }
  }];
};
